File: backend/app/api/customers.py
from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.orm import Session
from typing import Optional
from uuid import UUID

from app.core import get_db, get_current_user
from app.models import User
from app.schemas import (
    CustomerCreate, CustomerUpdate, CustomerResponse, CustomerListResponse
)
from app.services import CustomerService

router = APIRouter(prefix="/customers", tags=["Customers"])


@router.post("", response_model=CustomerResponse, status_code=status.HTTP_201_CREATED)
async def create_customer(
    customer_data: CustomerCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Create a new customer"""
    # Check if phone already exists
    existing = CustomerService.get_customer_by_phone(db, customer_data.phone)
    if existing:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Customer with this phone number already exists"
        )
    
    customer = CustomerService.create_customer(db, customer_data)
    return customer


@router.get("", response_model=CustomerListResponse)
async def list_customers(
    skip: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=100),
    search: Optional[str] = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """List all customers with filtering"""
    customers, total = CustomerService.list_customers(db, skip, limit, search)
    return CustomerListResponse(total=total, customers=customers)


@router.get("/{customer_id}", response_model=CustomerResponse)
async def get_customer(
    customer_id: UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get customer by ID"""
    customer = CustomerService.get_customer_by_id(db, customer_id)
    if not customer:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Customer not found"
        )
    return customer


@router.put("/{customer_id}", response_model=CustomerResponse)
async def update_customer(
    customer_id: UUID,
    customer_data: CustomerUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Update customer details"""
    customer = CustomerService.get_customer_by_id(db, customer_id)
    if not customer:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Customer not found"
        )
    
    # Update fields
    update_data = customer_data.model_dump(exclude_unset=True)
    for field, value in update_data.items():
        setattr(customer, field, value)
    
    db.commit()
    db.refresh(customer)
    return customer


@router.delete("/{customer_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_customer(
    customer_id: UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Delete a customer and all related data (contracts, payments)"""
    customer = CustomerService.get_customer_by_id(db, customer_id)
    if not customer:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Customer not found"
        )
    
    # Delete related contracts and payments
    from app.models import EMIContract, EMIPayment
    contracts = db.query(EMIContract).filter(EMIContract.customer_id == customer_id).all()
    for contract in contracts:
        db.query(EMIPayment).filter(EMIPayment.contract_id == contract.id).delete()
        db.delete(contract)
    
    db.delete(customer)
    db.commit()
    return None

File: backend-node/src/routes/customers.js
import { CustomerService } from '../services/deviceService.js';
import { EMIContract, EMIPayment } from '../models/index.js';
import { serializeCustomer } from '../utils/serializers.js';
import { httpError, requireUser } from '../middleware/auth.js';
import { ah } from './auth.js';

const UPDATABLE = [
  'full_name', 'phone', 'alternate_phone', 'email', 'address', 'city', 'state',
  'pincode', 'emergency_contact_name', 'emergency_contact_phone', 'emergency_contact_relation',
];

export default function registerCustomerRoutes(router) {
  router.post('/customers', requireUser, ah(async (req, res) => {
    const data = req.body || {};
    const existing = await CustomerService.getCustomerByPhone(data.phone);
    if (existing) throw httpError(400, 'Customer with this phone number already exists');
    const customer = await CustomerService.createCustomer(data);
    res.status(201).json(serializeCustomer(customer));
  }));

  router.get('/customers', requireUser, ah(async (req, res) => {
    const skip = Math.max(0, parseInt(req.query.skip, 10) || 0);
    let limit = parseInt(req.query.limit, 10);
    limit = Number.isNaN(limit) ? 50 : Math.min(100, Math.max(1, limit));
    const search = req.query.search || null;
    const [customers, total] = await CustomerService.listCustomers({ skip, limit, search });
    res.json({ total, customers: customers.map(serializeCustomer) });
  }));

  router.get('/customers/:customerId', requireUser, ah(async (req, res) => {
    const customer = await CustomerService.getCustomerById(req.params.customerId);
    if (!customer) throw httpError(404, 'Customer not found');
    res.json(serializeCustomer(customer));
  }));

  router.put('/customers/:customerId', requireUser, ah(async (req, res) => {
    const customer = await CustomerService.getCustomerById(req.params.customerId);
    if (!customer) throw httpError(404, 'Customer not found');
    const data = req.body || {};
    for (const field of UPDATABLE) {
      if (Object.prototype.hasOwnProperty.call(data, field)) customer[field] = data[field];
    }
    await customer.save();
    res.json(serializeCustomer(customer));
  }));

  router.delete('/customers/:customerId', requireUser, ah(async (req, res) => {
    const customer = await CustomerService.getCustomerById(req.params.customerId);
    if (!customer) throw httpError(404, 'Customer not found');
    const contracts = await EMIContract.findAll({ where: { customer_id: customer.id } });
    for (const c of contracts) {
      await EMIPayment.destroy({ where: { contract_id: c.id } });
      await c.destroy();
    }
    await customer.destroy();
    res.status(204).end();
  }));
}
